File: ios/Sources/Mekuri/MekuriFoldPass.swift
import CoreGraphics

/// What one fold render needs for a direction, progress and lifting corner.
/// The shader only folds a flap entering from the right with its free corner
/// at the top, so a right-to-left turn mirrors the page horizontally around
/// it and a bottom lift mirrors it vertically; `shaderProgress` is never
/// negative.
struct MekuriFoldPass: Equatable {
    let isMirrored: Bool
    let liftsFromBottom: Bool
    let shaderProgress: CGFloat

    init(direction: MekuriDirection, progress: CGFloat, liftsFromBottom: Bool = false) {
        self.isMirrored = direction == .rightToLeft
        self.liftsFromBottom = liftsFromBottom
        self.shaderProgress = abs(direction.sweep(progress: progress))
    }

    /// Horizontal scale applied on both sides of the layer effect.
    var mirrorScale: CGFloat { self.isMirrored ? -1 : 1 }

    /// Vertical scale applied on both sides of the layer effect.
    var verticalScale: CGFloat { self.liftsFromBottom ? -1 : 1 }
}

File: android/mekuri/src/main/kotlin/studio/anicanon/mekuri/MekuriFoldPass.kt
package studio.anicanon.mekuri

import kotlin.math.abs

/**
 * What one fold render needs for a direction, progress and lifting corner. The
 * shader only folds a flap entering from the right with its free corner at the
 * top, so a right-to-left turn mirrors the page horizontally around it and a
 * bottom lift mirrors it vertically; [shaderProgress] is never negative.
 */
internal data class MekuriFoldPass(
    val isMirrored: Boolean,
    val shaderProgress: Float,
    val liftsFromBottom: Boolean = false,
) {
    constructor(direction: MekuriDirection, progress: Float, liftsFromBottom: Boolean = false) : this(
        isMirrored = direction == MekuriDirection.RightToLeft,
        shaderProgress = abs(direction.sweep(progress)),
        liftsFromBottom = liftsFromBottom,
    )

    /** Horizontal scale applied on both sides of the fold effect. */
    val mirrorScale: Float
        get() = if (isMirrored) -1f else 1f

    /** Vertical scale applied on both sides of the fold effect. */
    val verticalScale: Float
        get() = if (liftsFromBottom) -1f else 1f

    internal companion object {
        /** The mirror depends on the direction alone, not on progress. */
        fun mirrorScale(direction: MekuriDirection): Float =
            MekuriFoldPass(direction, 0f).mirrorScale

        /** The vertical mirror depends on the lifting corner alone. */
        fun verticalScale(liftsFromBottom: Boolean): Float =
            MekuriFoldPass(MekuriDirection.LeftToRight, 0f, liftsFromBottom).verticalScale
    }
}
